### PR TITLE
Blog/Page/Post/Tag uri customisation

### PR DESCRIPTION
**Pull request by sbondaryev**
Add possibility to configure uri root directory for pages, posts and tags
via config params:
```
:page-root-uri
:post-root-uri
:tag-root-uri
````
- to place entity to the root directory use empty value (ex. :page-root-uri "")
- if parameter is not set corresponding {entity}-root parameter will be used
- set empty blog-prefix to use server root

Issues:
https://github.com/cryogen-project/cryogen-core/issues/32
https://github.com/cryogen-project/cryogen/issues/83

Related PR:
https://github.com/cryogen-project/cryogen/pull/96

### CODE REVIEW

**Comment by sbondaryev** [comment]
The core thing is the function io/path - which compose the path from elements and throw appart the empty ones (you can also specify the "/" on the beginning and the end .
Example:
(path "/" "blog" "post" "post.name")   -> "/blog/post/post.name"
(path "/" "post" "img" "/")    -> "/post/post/img/"

I also added the function io/create-file (similar to io/create-folder) to localise "public" variable in "io" namespace.

All the other changes are:
- replacements "str" with "path" to compose urls and file paths
- replacements "spit" with "create-file" to save files in public dir

File: src/cryogen_core/compiler.clj [hunk 1]
```diff
@@ -331,11 +330,11 @@
 
 (defn copy-resources-from-markup-folders
   "Copy resources from markup folders"
-  [config]
+  [{:keys [post-root page-root] :as config}]
   (copy-resources
     (merge config
            {:resources     (for [mu (m/markups)
-                                 t ["posts" "pages"]] (str (m/dir mu) "/" t))
+                                 t (distinct [post-root page-root])] (str (m/dir mu) "/" t))
```
Comment (sbondaryev) [inline src/cryogen_core/compiler.clj]: Replace hardcoded post/pages dir names with names from the config

File: src/cryogen_core/markup.clj [hunk 2]
```diff
@@ -16,7 +17,8 @@
 
   ex. <img src='/img/cryogen.png'/> becomes <img src='/blog/img/cryogen.png'/>"
   [blog-prefix text]
-  (clojure.string/replace text #"href=.?/|src=.?/" #(str (subs % 0 (dec (count %))) blog-prefix "/")))
+  (if-not (s/blank? blog-prefix)
```
Comment (sbondaryev) [inline src/cryogen_core/markup.clj]: Do nothing if blog-prefix is empty

**Comment by danielcompton** [comment]
Does this let you have URL's like `2015/11/23/mypost`?

**Comment by sbondaryev** [comment]
No, it does not.
But it could be a good issue to discuss this feature with project maintainers.
As it could affect quite a lot the current directores structure

**Comment by lacarmen** [comment]
Hey @sbondaryev, I finally got around to looking at this pr. Sorry it took so long! Everything looks great but there's one little bug. On line 293 of `compiler.clj` you have

```clojure
(path "/" blog-prefix "p" "index.html")
```

as the `else` condition but it should be 

```clojure
(path "/" blog-prefix "p" (str index ".html"))
```

since `index` is supposed to be the page number of each preview page. (I suppose it can be named better, you can change it if you want.)

If you could fix that and then [squash your commits](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) I'll merge in the pr. :)

**Comment by sbondaryev** [comment]
Hello, sure I will take a look this week-end

**Comment by sbondaryev** [comment]
done

**Comment by lacarmen** [comment]
Thanks! I'll push a new version to Clojars and update the template as well as the docs.

**Pull request merged**